modules/consentManagement.js

```diff
@@ -0,0 +1,164 @@
+const DEFAULT_ALLOW_AUCTION_WO_CONSENT = true;
+
+export let userCMP;
+export let consentTimeout;
+export let allowAuction;
+
+let consentData;
+
+let context;
+let args;
+let nextFn;
+
+let timer;
+let haveExited;
+
+// add new CMPs here, with their dedicated lookup function that passes the consentString to postLookup()
+const cmpCallMap = {
+  'appnexus': lookupAppNexusConsent
+};
+
+/**
+ * This function handles interacting with the AppNexus CMP to obtain the consentString value of the user.
+ * Given the asynch nature of the CMP's API, we pass in acting success/error callback functions to exit this function
+ * based on the appropriate result.
+ * @param {function(string)} cmpSuccess acts as a success callback when CMP returns a value; pass along consentString (string) from CMP
+ * @param {function(string)} cmpError acts as an error callback while interacting with CMP; pass along an error message (string)
+ */
+function lookupAppNexusConsent(cmpSuccess, cmpError) {
+  if (!window.__cmp) {
+    return cmpError('AppNexus CMP not detected.');
+  }
+
+  // first lookup - to determine if new or existing user
+  // if new user, then wait for user to make a choice and then run postLookup method
+  // if existing user, then skip to postLookup method
+  window.__cmp('getConsentData', 'vendorConsents', function(consentString) {
+    if (consentString == null) {
+      window.__cmp('addEventListener', 'onSubmit', function() {
+        // redo lookup to find new string based on user's choices
+        window.__cmp('getConsentData', 'vendorConsents', cmpSuccess);
+      });
+    } else {
+      cmpSuccess(consentString);
+    }
+  });
+}
+
+/**
+ * If consentManagement module is enabled (ie included in setConfig), this hook function will attempt to fetch the
+ * user's encoded consent string from the supported CMP.  Once obtained, the module will store this
+ * data as part of a gdprConsent object and transferred to adaptermanager's gdprDataHandler object.
+ * This information is later added into the bidRequest object for any supported adapters to read/pass along to their system.
+ * @param {object} config This is the same param that's used in pbjs.requestBids.  The config.adunits will be updated.
+ * @param {function} fn The next function in the chain, used by hook.js
+ */
+export function requestBidsHook(config, fn) {
+  context = this;
+  args = arguments;
+  nextFn = fn;
+  haveExited = false;
+
+  // in case we already have consent (eg during bid refresh)
+  if (consentData) {
+    return nextFn.apply(context, args);
+  }
+
+  if (!Object.keys(cmpCallMap).includes(userCMP)) {
+    utils.logWarn(`CMP framework (${userCMP}) is not a supported framework.  Aborting consentManagement module and resuming auction.`);
+    return nextFn.apply(context, args);
+  }
+
+  // lookup times and user interaction with CMP prompts can greatly vary, so enforcing a timeout on the CMP process
+  timer = setTimeout(cmpTimedOut, consentTimeout);
```

Comment:
Please skip the setTimeout if `consentTimeout` is 0. Otherwise we lose out place in the javascript stack.

Reply:
@mkendall07 Talked about this change with me this morning.  I've got the changes put together and will include them in the next commit.

As a confirmation question - if the publisher were to have the setting to cancel the auction enabled, the combination of `consentTimeout=0` + the cancel setting would cause a new user's auction to be canceled immediately during that first page load where they have to fill out the consent information.  

Any subsequent page loads for that new user would go through the normal workflow and should be fine (even if the publisher had the `consentTimeout` at 0, but is that first time canceled auction an issue?  Or is it fine to leave as is?

Reply:
First time cancelation is fine -- it's a pretty straightforward implication of the publisher's wishes in this scenario, IMO.